pkg/kubelet/container/helpers.go

```diff
@@ -96,7 +97,13 @@ func ShouldContainerBeRestarted(container *v1.Container, pod *v1.Pod, podStatus
 // the running container with its desired spec.
 func HashContainer(container *v1.Container) uint64 {
 	hash := fnv.New32a()
-	hashutil.DeepHashObject(hash, *container)
+	// Omit nil or empty field when calculating hash value
+	// Please see https://github.com/kubernetes/kubernetes/issues/53644
+	oldJsonBytes, _ := json.Marshal(container)
+	newObj := make(map[string]interface{})
+	json.Unmarshal([]byte(oldJsonBytes), &newObj)
```

Comment:
No need to unmarshal - just hash the marshalled json directly:
```
err := json.NewEncoder(hash).Encode(container)
```
Actually, I think this is required to fix the issue.

Reply:
@tallclair This is simpler than mine. Thanks.

Some comments from @liggitt on this kind of change,
> this change will still not prevent a new defaulted field from changing the hash, and seems like it would guarantee that hashes would change between 1.9 and 1.10, which isn't great.

@tallclair @liggitt What about the 2nd option in [my above comment](https://github.com/kubernetes/kubernetes/pull/57741#issuecomment-356196638)?

Reply:
Oh yeah, I should have read the past discussion before reviewing... How / where / when is this hash value used? Would an option to skip defaulting fix it? I didn't follow your suggestion.

Reply:
defaulted values happen on the API server side, so we can't really prevent them here. we have tests in place now that alert us if a field under Pod is defaulted

in a release where we have a new field introduced to podspec (like we do in 1.16), the hashes would change anyway, so it's a reasonable release to add this

Reply:
I'd like to see this happen in 1.16